README.md
# 🃏JOKER-ENHANCER⚡

🃏 An Overall User Experience Enhancer with Many Tweaks or Enhancements ⚡

🟢 Do you feel laggy in your ROM?
  ✅ Then, Try Smoothness Enhancement

🟢 Do you want latest security patch but don't want to change the ROM?
  ✅ Then, Try Latest Security Patch

🟢 Do you like Amoled Colors?
  ✅ Then, Try Amoled Like Colors 

⏩ And Many More waiting for you

🃏 Feel The Joker's Supremacy 🔮

 


changelog.md
### ARCEUS - HOTFIX - 21-04-2023
* Fixed System Props Related Things
* Fixed Joker AI & Removed Alarm
* Fixed Recovery Uninstallation
* Fixed Auto Reboot
* Fixed Bootloop saver (May not work with magisk delta)
* Rebased Module
* Fixed Debloater


JOKER/tweaks/NRT.sh
# Works without rebooting device as well

# Write Function
write() {

 [[ ! -f "$1" ]] && return 1

 chmod +w "$1" 2> /dev/null

 if ! echo "$2" > "$1"   2> /dev/null
 then
  echo "Failed: $1 → $2" > /dev/null 2>&1
  return 1
 fi

 echo " SUCCESS $1 → $2" > /dev/null 2>&1
}

# Optimization 
OPT() {
sleep 1
cmd package bg-dexopt-job
fstrim -v /cache
sleep 0.5
fstrim -v /data
sleep 0.5
fstrim -v /product
sleep 0.5
fstrim -v /system
sleep 0.5
fstrim -v /vendor
sleep 0.4
   }
   
# Amoled Colors 
AMO() {
   write "/sys/module/msm_drm/parameters/kcal_cont" "262"
write "/sys/module/msm_drm/parameters/kcal_hue" "0"
write "/sys/module/msm_drm/parameters/kcal_sat" "272"
write "/sys/module/msm_drm/parameters/kcal_val" "256"
   }

# Smoothness 
SMO() {
write "/dev/stune/top-app/schedtune.boost" "10"
write "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor" "schedutil"
write "/sys/devices/system/cpu/cpu6/cpufreq/scaling_governor" "schedutil"
write "/sys/module/cpu_boost/parameters/input_boost_freq" "0:1708800 6:2208800"
write "/sys/module/cpu_boost/parameters/input_boost_ms" "500"
write "/sys/class/kgsl/kgsl-3d0/min_pwrlevel" "4"
write "/proc/sys/kernel/sched_downmigrate" "78"
write "/proc/sys/kernel/sched_upmigrate" "93"
 }

# Network Tweaks
NET() {
write "/proc/sys/net/ipv4/tcp_ecn" "1"
write "/proc/sys/net/ipv4/tcp_fastopen" "8"
write "/proc/sys/net/ipv4/tcp_syncookies" "0"
}


JOKER/scripts/functions.sh
# Joker Functions

# Read Any Value
READ() {
  value=$(sed -e '/^[[:blank:]]*#/d;s/[\t\n\r ]//g;/^$/d' "$2" | grep -m 1 "^$1=" | cut -d'=' -f 2)
  echo "$value"
  return $?
}

# What to do after Aborting?
abort() {
  ui_print "$1"
  rm -rf $MODPATH 2>/dev/null
  cleanup
  rm -rf $TMPDIR 2>/dev/null
  exit 1
}

# Cleans Module Directory 
cleanup() {
  rm -rf $MODPATH/JOKER 2>/dev/null
}

# Check Device 
device_check() {
  local opt=`getopt -o dm -- "$@"` type=device
  eval set -- "$opt"
  while true; do
    case "$1" in
      -d) local type=device; shift;;
      -m) local type=manufacturer; shift;;
      --) shift; break;;
      *) abort "Invalid device_check argument $1! Aborting!";;
    esac
  done
  local prop=$(echo "$1" | tr '[:upper:]' '[:lower:]')
  for i in /system /vendor /odm /product; do
    if [ -f $i/build.prop ]; then
      for j in "ro.product.$type" "ro.build.$type" "ro.product.vendor.$type" "ro.vendor.product.$type"; do
        [ "$(sed -n "s/^$j=//p" $i/build.prop 2>/dev/null | head -n 1 | tr '[:upper:]' '[:lower:]')" == "$prop" ] && return 0
      done
    fi
  done
  return 1
}

# Copy Paste
cp_ch() {
  local opt=`getopt -o nr -- "$@"` BAK=true UBAK=true FOL=false
  eval set -- "$opt"
  while true; do
    case "$1" in
      -n) UBAK=false; shift;;
      -r) FOL=true; shift;;
      --) shift; break;;
      *) abort "Invalid cp_ch argument $1! Aborting!";;
    esac
  done
  local SRC="$1" DEST="$2" OFILES="$1"
  $FOL && local OFILES=$(find $SRC -type f 2>/dev/null)
  [ -z $3 ] && PERM=0644 || PERM=$3
  case "$DEST" in
    $TMPDIR/*|$MODULEROOT/*|$NVBASE/modules/$MODID/*) BAK=false;;
  esac
  for OFILE in ${OFILES}; do
    if $FOL; then
      if [ "$(basename $SRC)" == "$(basename $DEST)" ]; then
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST|")
      else
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST/$(basename $SRC)|")
      fi
    else
      [ -d "$DEST" ] && local FILE="$DEST/$(basename $SRC)" || local FILE="$DEST"
    fi
    if $BAK && $UBAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
      [ -f "$FILE" -a ! -f "$FILE~" ] && { mv -f $FILE $FILE~; echo "$FILE~" >> $INFO; }
    elif $BAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
    fi
    install -D -m $PERM "$OFILE" "$FILE"
  done
}

# Installation Script
install_script() {
  case "$1" in
    -l) shift; local INPATH=$NVBASE/service.d;;
    -p) shift; local INPATH=$NVBASE/post-fs-data.d;;
    *) local INPATH=$NVBASE/service.d;;
  esac
  [ "$(grep "#!/system/bin/sh" $1)" ] || sed -i "1i #!/system/bin/sh" $1
  local i; for i in "MODPATH" "LIBDIR" "MODID" "INFO" "MODDIR"; do
    case $i in
      "MODPATH") sed -i "1a $i=$NVBASE/modules/$MODID" $1;;
      "MODDIR") sed -i "1a $i=\${0%/*}" $1;;
      *) sed -i "1a $i=$(eval echo \$$i)" $1;;
    esac
  done
  [ "$1" == "$MODPATH/uninstall.sh" ] && return 0
  case $(basename $1) in
    post-fs-data.sh|service.sh) ;;
    *) cp_ch -n $1 $INPATH/$(basename $1) 0755;;
  esac
}

# Prop Process
prop_process() {
  sed -i -e "/^#/d" -e "/^ *$/d" $1
  [ -f $MODPATH/system.prop ] || mktouch $MODPATH/system.prop
  while read LINE; do
    echo "$LINE" >> $MODPATH/system.prop
  done < $1
}

# Check for min/max api version
[ -z $MINAPI ] || { [ $API -lt $MINAPI ] && abort "! Your system API of $API is less than the minimum api of $MINAPI! Aborting!"; }
[ -z $MAXAPI ] || { [ $API -gt $MAXAPI ] && abort "! Your system API of $API is greater than the maximum api of $MAXAPI! Aborting!"; }

# Set variables
[ $API -lt 26 ] && DYNLIB=false
[ -z $DYNLIB ] && DYNLIB=false
[ -z $DEBUG ] && DEBUG=false
[ -e "$PERSISTDIR" ] && PERSISTMOD=$PERSISTDIR/magisk/$MODID
INFO=$NVBASE/modules/.$MODID-files
ORIGDIR="$MAGISKTMP/mirror"
if $DYNLIB; then
  LIBPATCH="\/vendor"
  LIBDIR=/system/vendor
else
  LIBPATCH="\/system"
  LIBDIR=/system
fi
if ! $BOOTMODE; then
  ui_print "- Only uninstall is supported in recovery"
  ui_print "  Uninstalling!"
  touch $MODPATH/remove
  [ -s $INFO ] && install_script $MODPATH/uninstall.sh || rm -f $INFO $MODPATH/uninstall.sh
  recovery_cleanup
  cleanup
  rm -rf $NVBASE/modules_update/$MODID $TMPDIR 2>/dev/null
  exit 0
fi

# Debug
if $DEBUG; then
  ui_print " - Debug Mode Activated "
  ui_print " - Module install log will include debug info"
  ui_print " - Be sure to save it after module install"
  set -x
fi

# Extract files
unzip -o "$ZIPFILE" -x 'META-INF/*' 'JOKER/functions.sh' -d $MODPATH >&2
[ -f "$MODPATH/JOKER/addon.tar.xz" ] && tar -xf $MODPATH/JOKER/addon.tar.xz -C $MODPATH/JOKER 2>/dev/null

# Run addons
if [ "$(ls -A $MODPATH/JOKER/addon/*/install.sh 2>/dev/null)" ]; then
  ui_print " "; ui_print "- Running Addons -"
  for i in $MODPATH/JOKER/addon/*/install.sh; do
    ui_print "  Running $(echo $i | sed -r "s|$MODPATH/JOKER/addon/(.*)/install.sh|\1|")..."
    . $i
  done
fi

# Remove files outside of module directory
if [ -f $INFO ]; then
  while read LINE; do
    if [ "$(echo -n $LINE | tail -c 1)" == "~" ]; then
      continue
    elif [ -f "$LINE~" ]; then
      mv -f $LINE~ $LINE
    else
      rm -f $LINE
      while true; do
        LINE=$(dirname $LINE)
        [ "$(ls -A $LINE 2>/dev/null)" ] && break 1 || rm -rf $LINE
      done
    fi
  done < $INFO
  rm -f $INFO
fi

### Install
[ -f "$MODPATH/joker.sh" ] && . $MODPATH/joker.sh

# Remove comments from files and place them, add blank line to end if not already present
for i in $(find $MODPATH -type f -name "*.sh" -o -name "*.prop" -o -name "*.rule"); do
  [ -f $i ] && { sed -i -e "/^#/d" -e "/^ *$/d" $i; [ "$(tail -1 $i)" ] && echo "" >> $i; } || continue
  case $i in
    "$MODPATH/service.sh") install_script -l $i;;
    "$MODPATH/post-fs-data.sh") install_script -p $i;;
    "$MODPATH/uninstall.sh") if [ -s $INFO ] || [ "$(head -n1 $MODPATH/uninstall.sh)" != "# Don't modify anything after this" ]; then
                               install_script $MODPATH/uninstall.sh
                             else
                               rm -f $INFO $MODPATH/uninstall.sh
                             fi;;
  esac
done

$IS64BIT || for i in $(find $MODPATH/system -type d -name "lib64"); do rm -rf $i 2>/dev/null; done  
[ -d "/system/priv-app" ] || mv -f $MODPATH/system/priv-app $MODPATH/system/app 2>/dev/null
[ -d "/system/xbin" ] || mv -f $MODPATH/system/xbin $MODPATH/system/bin 2>/dev/null
if $DYNLIB; then
  for FILE in $(find $MODPATH/system/lib* -type f 2>/dev/null | sed "s|$MODPATH/system/||"); do
    [ -s $MODPATH/system/$FILE ] || continue
    case $FILE in
      lib*/modules/*) continue;;
    esac
    mkdir -p $(dirname $MODPATH/system/vendor/$FILE)
    mv -f $MODPATH/system/$FILE $MODPATH/system/vendor/$FILE
    [ "$(ls -A `dirname $MODPATH/system/$FILE`)" ] || rm -rf `dirname $MODPATH/system/$FILE`
  done
  # Delete empty lib folders (busybox find doesn't have this capability)
  toybox find $MODPATH/system/lib* -type d -empty -delete >/dev/null 2>&1
fi

# Set permissions
set_perm_recursive $MODPATH 0 0 0755 0644
if [ -d $MODPATH/system/vendor ]; then
  set_perm_recursive $MODPATH/system/vendor 0 0 0755 0644 u:object_r:vendor_file:s0
  [ -d $MODPATH/system/vendor/app ] && set_perm_recursive $MODPATH/system/vendor/app 0 0 0755 0644 u:object_r:vendor_app_file:s0
  [ -d $MODPATH/system/vendor/etc ] && set_perm_recursive $MODPATH/system/vendor/etc 0 0 0755 0644 u:object_r:vendor_configs_file:s0
  [ -d $MODPATH/system/vendor/overlay ] && set_perm_recursive $MODPATH/system/vendor/overlay 0 0 0755 0644 u:object_r:vendor_overlay_file:s0
  for FILE in $(find $MODPATH/system/vendor -type f -name *".apk"); do
    [ -f $FILE ] && chcon u:object_r:vendor_app_file:s0 $FILE
  done
fi
set_permissions

# Complete install
cleanup


JOKER/scripts/post-fs-data.sh
# Module Path 
MODDIR=${0%/*}

# Write function to avoid any error
write() {

 [[ ! -f "$1" ]] && return 1

 chmod +w "$1" 2> /dev/null

 if ! echo "$2" > "$1"   2> /dev/null
 then
  echo "Failed: $1 → $2" > /dev/null 2>&1
  return 1
 fi

 echo " SUCCESS $1 → $2" > /dev/null 2>&1
}

# Feel The Smoothness
smt() {
write "/dev/stune/top-app/schedtune.boost" "10"
write "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor" "schedutil"
write "/sys/devices/system/cpu/cpu6/cpufreq/scaling_governor" "schedutil"
write "/sys/module/cpu_boost/parameters/input_boost_freq" "0:1708800 6:2208800"
write "/sys/module/cpu_boost/parameters/input_boost_ms" "500"
write "/sys/class/kgsl/kgsl-3d0/min_pwrlevel" "4"
write "/proc/sys/kernel/sched_downmigrate" "78"
write "/proc/sys/kernel/sched_upmigrate" "93"
 }

# Amoled Like Colors
alc() {
write "/sys/module/msm_drm/parameters/kcal_cont" "262"
write "/sys/module/msm_drm/parameters/kcal_hue" "0"
write "/sys/module/msm_drm/parameters/kcal_sat" "272"
write "/sys/module/msm_drm/parameters/kcal_val" "256"
    }

# Improve Ram Management   
ram() {
write "/sys/block/zram0/reset" "1"
write "/sys/block/zram0/disksize" "1894436864"
write "/sys/block/zram0/max_comp_streams" "8"
write "/proc/sys/vm/swappiness" "100"
write "/proc/sys/vm/dirty_background_ratio" "15"
write "/proc/sys/vm/vfs_cache_pressure" "200"
write "/proc/sys/vm/dirty_writeback_centisecs" "3000"
write "/proc/sys/vm/drop_caches" "3"
write "/proc/sys/vm/oom_kill_allocating_task" "0"
write "/sys/module/lowmemorykiller/parameters/minfree" "256,10240,32000,34000,36000,38000"
}

# Network Tweaks
nt() {
write "/proc/sys/net/ipv4/tcp_ecn" "1"
write "/proc/sys/net/ipv4/tcp_fastopen" "8"
write "/proc/sys/net/ipv4/tcp_syncookies" "0"
}

# Applied Tweaks


install.sh
# Variables
PROPFILE=true
POSTFSDATA=true
LATESTARTSERVICE=true
SKIPMOUNT=false
DEBUG=false

# Replace list
REPLACE="
"

# Defines
source /$TMPDIR/functions.sh
config=/sdcard/joker.config
Allow=$(READ "Allow" "$config")
ConfigVersion=$(READ "ConfigVersion" "$config")
Optimization=$(READ "Optimization" "$config")
AmoledColors=$(READ "AmoledColors" "$config")
Network=$(READ "Network" "$config")
Smoothness=$(READ "Smoothness" "$config")
SecurityPatch=$(READ "SecurityPatch" "$config")
Ram=$(READ "Ram" "$config")

# Display While Installing Module
ui_print ""
ui_print "          🃏 JOKER ENHANCER ⚡"
ui_print " "
ui_print "          ℹ️ Version = $MODVERSION 🔥"
ui_print " "
ui_print " ⚠️ REMOVE ANY OTHER ENHANCER OR TWEAKER MODULE ⚠️"
sleep 1

# Custom Script For Installation
on_install() {
unzip -o "$ZIPFILE" 'JOKER/apps/*' -d $TMPDIR >&2
unzip -o "$ZIPFILE" 'JOKER/files/*' -d $TMPDIR >&2
unzip -o "$ZIPFILE" 'JOKER/scripts/*' -d $TMPDIR >&2
unzip -o "$ZIPFILE" 'JOKER/tweaks/*' -d $TMPDIR >&2
  unzip -o "$ZIPFILE" 'JOKER/scripts/functions.sh' -d $TMPDIR >&2  
  unzip -o "$ZIPFILE" -x 'META-INF/*' 'JOKER/scripts/functions.sh' 'install.sh' -d $MODPATH >&2
source /$TMPDIR/JOKER/tweaks/NRT.sh
AR=$TMPDIR/service.sh
BR=$TMPDIR/post-fs-data.sh

sleep 1
## Real Installation Begin
if [ -f "$config" ]; then
ui_print ""
ui_print " 🔎 Found JOKER 🃏 Config"

sleep 1
# Abort if Config is OLD
if [ "$ConfigVersion" != "$MODVERSION" ]; then
ui_print ""
ui_print " ❎ Config is Old"
ui_print ""
ui_print "  🃏 Updating Config Version to $MODVERSION"
sed -i "/ConfigVersion/s/.*/ConfigVersion=$MODVERSION/" $TMPDIR/JOKER/files/joker.config
cp -r /$TMPDIR/JOKER/files/joker.config /sdcard/joker.config
ui_print ""
ui_print "   🔴 Edit According to your Needs"
ui_print ""
ui_print "    🔃 Reinstall 🃏 JOKER 🃏 After editing Config "
abort ""
fi

sleep 1
# Abort if Config not Edited or Not Given Permission To Joker 
if [ "$Allow" = "0" ] ; then
ui_print ""
ui_print " ❎ Config is not Edited"
ui_print ""
ui_print "  🃏 Edit Config File located in /sdcard/joker.config OR Internal Storage"
ui_print ""
ui_print "   ✔️ Change Allow=1 in joker.config and edit according to your choices"
ui_print ""
ui_print "    🔃 Reinstall 🃏 JOKER 🃏 After editing Config "
abort ""
fi

sleep 1
## No Reboot Required Tweaks
ui_print ""
ui_print " ☑️ Applying No Reboot Required Tweaks"
ui_print ""
ui_print "  ⚠️ Note :- Reboot is Required For Disabling Tweaks"

sleep 1
 # Optimization 
	if [ "$Optimization" = "1" ] ; then
    ui_print ""
    ui_print "  ✅ OPTIMIZING YOUR DEVICE "
     ui_print ""
     ui_print "   ℹ️ This may take few minutes "
     OPT
        else
     ui_print ""
     ui_print "    ⏩ SKIPPING OPTIMIZATION..."
fi

sleep 1
# Amoled Tweaks 
	if [ "$AmoledColors" = "1" ]; then
    ui_print ""
    ui_print "  ✅ ENABLING AMOLED LIKE COLORS "
    ui_print ""
    ui_print "   ℹ️ Turn on/off night light if doesn't get applied"
    AMO 
    echo "alc" >> $AR;
    echo "alc" >> $BR;
	else 
	ui_print ""
	ui_print "    ⏩ SKIPPING AMOLED LIKE COLORS..."
fi
 
 sleep 1
# Smoothness 
	if [ "$Smoothness" = "1" ] ; then
    ui_print ""
    ui_print "  ✅ ENHANCING SMOOTHNESS"
     ui_print ""
     ui_print "   ℹ️ Recommended for A12 - A13 Specially "
        SMO
    echo "smt" >> $AR;
    echo "smt" >> $BR;
else
     ui_print ""
     ui_print "    ⏩ SKIPPING OPTIMIZATION..."
fi

sleep 1
# Network Tweaks For JIO Unlimited 
	if [ "$Network" = "1" ] ; then
    ui_print ""
    ui_print "  ✅ TWEAKING NETWORK FOR JIO UNLIMITED"
     ui_print ""
     ui_print "   ⏺️ Turn on/off airplane mode if doesn't work "
     NET
    echo "nt" >> $AR;
    echo "nt" >> $BR;
        else
     ui_print ""
     ui_print "    ⏩ SKIPPING NETWORK TWEAK ENHANCEMENT..."
fi

sleep 1
## Reboot Required Tweaks
ui_print ""
ui_print " ☑️ Applying Reboot Required Tweaks"

sleep 1
# Latest Security Patch Level
	if [ "$SecurityPatch" = "1" ] ; then
    ui_print ""
    ui_print "  ✅ UPDATING SECURITY PATCH"
    SPL
     ui_print ""
     ui_print "   ⚠️ May Break Safety Net"
        else
     ui_print ""
     ui_print "    ⏩ SKIPPING SECURITY PATCH ENHANCEMENT..."
fi

sleep 1
# RAM
	if [ "$Ram" = "1" ] ; then
    ui_print ""
    ui_print "  ✅ ENHANCING RAM MANAGEMENT"
    RAME
     ui_print ""
     ui_print "   ℹ️ It Will change RAM behaviour to less killing apps"
     echo "ram" >> $AR;
     echo "ram" >> $BR;
     sed -i "/ro.build.version.security_patch/s/.*/ro.build.version.security_patch=2023-02-05/" $TMPDIR/system.prop
sed -i "/ro.vendor.build.security_patch/s/.*/ro.vendor.build.security_patch=2023-02-05/" $TMPDIR/system.prop
        else
     ui_print ""
     ui_print "    ⏩ SKIPPING RAM MANAGEMENT ENHANCEMENT..."
fi

ui_print ""
ui_print "           ✅ INSTALLATION DONE ✅"
ui_print ""
ui_print "           🃏 FEEL THE JOKER'S SUPREMACY 🃏"
ui_print ""

sleep 1
# Abort If Config Not Found
else 
ui_print ""
ui_print " ❎ No Joker Config Found"
ui_print ""
ui_print "  🃏 Making One Default Config in /sdcard/joker.config"
cp -r /$TMPDIR/JOKER/files/joker.config /sdcard/joker.config
ui_print ""
ui_print "   🔴 Edit According to your Needs then comeback"
ui_print ""
ui_print "    🔃 Reinstall 🃏 JOKER 🃏 After editing config "
abort ""
fi
}

# Set Necessary Permissions
set_permissions() {
  set_perm_recursive $MODPATH 0 0 0755 0644
 set_perm_recursive  $MODPATH/system/lib       0     0       0755      0644
   set_perm  $MODPATH/system/bin/app_process32   0     2000    0755      u:object_r:zygote_exec:s0
  set_perm  $MODPATH/system/bin/dex2oat         0     2000    0755      u:object_r:dex2oat_exec:s0
   set_perm  $MODPATH/system/lib/libart.so       0     0       0644
}
